### ISSUE TITLE
Provide `arm64` build of `tsc-dyn.dylib` for Apple Silicon (M1) Macs

### ISSUE THREAD

**Issue by 3c1u** [emacs-tree-sitter/elisp-tree-sitter#88]
Currently, `tsc-dyn.dylib` has only `x86_64` binary.

```
✗ ~ file tsc-dyn.dylib 
tsc-dyn.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

`tsc-dyn.dylib` can be cross-compiled using the latest Rust toolchain.

**Comment by ubolonton** [comment]
This will likely need system toolchain for arm64 as well, in addition to Rust toolchain.
Then we will need someone with an M1 Mac to test it.

Have you been able to cross-compile it? Or do you have a link to a working guide for targeting `aarch64-apple-darwin` on `x86_64-apple-darwin`?

**Comment by 3c1u** [comment]
Tested on x86_64 macOS 11.1.

Use `master` branch of `tree-sitter` (`arm64` is now supported), and `emacs` crate with updated `bindgen` (libclang fails to load somehow otherwise)
```toml
emacs = { path = "../../emacs-module-rs" }
tree-sitter = { git = "https://github.com/tree-sitter/tree-sitter.git" }
```

and run this command:
```
env SDKROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk" cargo build --all --release --target aarch64-apple-darwin
```

**Comment by ubolonton** [comment]
Did you install XCode Beta?

My MacOS SDK doesn't support `arm64`. I'm getting `error architecture not supported`.
I don't have access to XCode Beta, so I cannot try this out at the moment.

https://stackoverflow.com/questions/64313634/xcode-12-compiling-for-macos-arm64-arch

**Comment by 3c1u** [comment]
Nope, I'm using the latest stable Xcode 12.3 (12C33) on macOS Big Sur 11.1.

**Comment by ubolonton** [comment]
Ah ok, I misread as 10.11. Yeah, Big Sur's SDK supports `arm64` arch.

I'll check this again when I have access to an SDK supporting `arm64`.

**Comment by ubolonton** [comment]
> and `emacs` crate with updated `bindgen` (libclang fails to load somehow otherwise)

Does it work if you delete `Cargo.lock` instead?

**Comment by 3c1u** [comment]
Deleting `Cargo.lock` worked on x86_64 cross-compilation. I also tried on an Apple Silicon Mac and libclang fails to load, which cannot be solved by deleting `Cargo.lock`.

**Comment by theothornhill** [comment]
I'm struggling with how to build this on my M1 mac. I seem to be stuck on resolving dependencies, the earliest error being:

```
   Compiling emacs_module v0.12.0
The following warnings were emitted during compilation:

warning: couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
warning: set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

error: failed to run custom build command for `emacs_module v0.12.0`

Caused by:
  process didn't exit successfully: `/Users/theodor/Git/emacs-tree-sitter/target/debug/build/emacs_module-e1ade98717ceb887/build-script-build` (exit code: 101)
  --- stdout
  cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
  cargo:warning=set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

  --- stderr
  thread 'main' panicked at 'libclang error; possible causes include:
  - Invalid flag syntax
  - Unrecognized flags
  - Invalid flag arguments
  - File I/O errors
  If you encounter an error missing from this list, please file an issue or a PR!', /Users/theodor/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.51.1/src/ir/context.rs:571:15
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

With some help from someone more knowledgeable in rust I can try to build this

**Comment by ubolonton** [comment]
@theothornhill Seems like you need to install `llvm`. It's needed to generate the Rust binding for `emacs-module.h`

**Comment by theothornhill** [comment]
Yeah, you're right. `emacs_module` fails on `bindgen-0.51`, as mentioned other places.

**Comment by theothornhill** [comment]
I think I managed to compile it to M1 now. 

running `file` on the created dylib returns
```
tsc-dyn.dylib: Mach-O 64-bit dynamically linked shared library arm64
```

However, now I'm stuck at compiling the `rust.dylib`.
Running `make ensure/rust` it fails with:
```
Debugger entered--Lisp error: (tsc-lang-load-failed "dlopen(/Users/theodor/Git/emacs-tree-sitter/langs/...")
  tsc--load-language("/Users/theodor/Git/emacs-tree-sitter/langs/bin/rus..." "tree_sitter_rust" rust)
  tree-sitter-load(rust nil nil)
  tree-sitter-require(rust)
  (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol)))
  (unwind-protect (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol))) (tree-sitter-langs--copy-query lang-symbol))
  tree-sitter-langs-ensure(rust)
  (progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'rust))
  eval((progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'rust)) t)
  command-line-1(("--directory" "/Users/theodor/Git/emacs-tree-sitter/core" "--directory" "/Users/theodor/Git/emacs-tree-sitter/lisp" "--directory" "/Users/theodor/Git/emacs-tree-sitter/langs" "--eval" "(progn (require 'tree-sitter-langs) (tree-sitter-l..."))
  command-line()
  normal-top-level()
```

after it has run the `tree-sitter test`. I've tried it both from normal terminal and shell inside emacs with load-path explicitly set before running the command. Any tips for how to proceed?

**Comment by douglasdavis** [comment]
I'm seeing something similar. I built `tsc-dyn.dylib` but I see the following when running `make ensure/python`:
```
$ EMACS=/Applications/Emacs.app/Contents/MacOS/Emacs make ensure/python
[...]
Debugger entered--Lisp error: (tsc-lang-abi-too-new 13 (9 . 12) "/Users/ddavis/software/repos/emacs-tree-sitter/lan...")
  tsc--load-language("/Users/ddavis/software/repos/emacs-tree-sitter/lan..." "tree_sitter_python" python)
  (let ((language (tsc--load-language full-path native-symbol-name lang-symbol))) (let* ((key lang-symbol)) (condition-case nil (with-no-warnings (map-put! tree-sitter-languages key language nil)) (map-not-inplace (setq tree-sitter-languages (map-insert tree-sitter-languages key language))))) language)
  (let* ((lang-name (symbol-name lang-symbol)) (fallback-name (replace-regexp-in-string "-" "_" lang-name)) (native-symbol-name (or native-symbol-name (format "tree_sitter_%s" fallback-name))) (files (if file (list file) (cons lang-name (if (string= lang-name fallback-name) nil (list fallback-name))))) (full-path (seq-some #'(lambda (base-name) (locate-file base-name tree-sitter-load-path tree-sitter-load-suffixes)) files))) (if full-path nil (error "Cannot find shared library for language: %S" lang-symbol)) (let ((language (tsc--load-language full-path native-symbol-name lang-symbol))) (let* ((key lang-symbol)) (condition-case nil (with-no-warnings (map-put! tree-sitter-languages key language nil)) (map-not-inplace (setq tree-sitter-languages (map-insert tree-sitter-languages key language))))) language))
  tree-sitter-load(python nil nil)
  (or (alist-get lang-symbol tree-sitter-languages) (tree-sitter-load lang-symbol file native-symbol-name))
  tree-sitter-require(python)
  (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol)))
  (unwind-protect (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol))) (tree-sitter-langs--copy-query lang-symbol))
  tree-sitter-langs-ensure(python)
  (progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'python))
  eval((progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'python)) t)
  command-line-1(("--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/cor..." "--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/lis..." "--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/lan..." "--eval" "(progn (require 'tree-sitter-langs) (tree-sitter-l..."))
  command-line()
  normal-top-level()

make: *** [ensure/python] Error 255
```

**Comment by ubolonton** [comment]
> `Debugger entered--Lisp error: (tsc-lang-load-failed "dlopen(/Users/theodor/Git/emacs-tree-sitter/langs/...")`

Emacs collapses the error message. You can click on the dots `...` to expand it. You can also modify/advise `tree-sitter-load` to catch that signal and print the full error data.

> `Debugger entered--Lisp error: (tsc-lang-abi-too-new 13 (9 . 12) "/Users/ddavis/software/repos/emacs-tree-sitter/lan...")`

This is a different error. It means that your `python.dylib` is too new. It was probably compiled by `tree-sitter` CLI version `0.19`. If you compile it yourself, you can try installing an older version of the CLI, e.g. `0.18`.

**Comment by douglasdavis** [comment]
> This is a different error. It means that your `python.dylib` is too new (13). It was probably compiled by `tree-sitter` CLI version `0.19`. If you compile it yourself, you can try installing an older version of the CLI, e.g. `0.18`.

Compiling the `tree-sitter` CLI at the 0.18.0 tag indeed worked. Thanks!

**Comment by tsujp** [comment]
I don't really know the ins and outs of tree-sitter intimately, or elisp as I'm still new to both of these but I played around with this all for an hour or two and got as far as manually compiling a parser for arm64 (using gcc not clang) but emacs segfaults when that's done. I've looked at the source for `emacs-tree-sitter` as well as `tree-sitter-langs` and `tree-sitter-c-sharp` and as far as I understand it the `parser.c` is generated from [here](https://github.com/ubolonton/tree-sitter-langs/blob/master/tree-sitter-langs-build.el#L283). Where this is compiled into a binary I don't know. I looked through all 3 repos for "clang" and "ld" (which are the things being used as I mucked around a bit and ended up getting this out)

```
Parser compilation failed.
Stdout: 
Stderr: Undefined symbols for architecture x86_64:
  "_tree_sitter_c_sharp_external_scanner_create", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_deserialize", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_destroy", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_scan", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_serialize", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Error calling (tree-sitter test), exit code is 1
make: *** [Makefile:13: ensure/c-sharp] Error 255
```

So I've hit a dead end.

**Comment by theothornhill** [comment]
>Emacs collapses the error message. You can click on the dots ... to expand it. You can also modify/advise tree-sitter-load to catch that signal and print the full error data.

The error was just that `c-sharp.dylib` was wrong architecture. I tried to manually run `tree-sitter generate`

Now I'm also stuck on what looks like the same issue as above.

running `clang -o parser.so -shared src/parser.c -I./src -v`

```
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: arm64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/clang" -cc1 -triple arm64-apple-macosx11.0.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name parser.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mframe-pointer=non-leaf -fno-strict-return -masm-verbose -munwind-tables -target-sdk-version=11.1 -target-cpu vortex -target-feature +v8.3a -target-feature +fp-armv8 -target-feature +neon -target-feature +crc -target-feature +crypto -target-feature +fullfp16 -target-feature +ras -target-feature +lse -target-feature +rdm -target-feature +rcpc -target-feature +zcm -target-feature +zcz -target-feature +sha2 -target-feature +aes -target-abi darwinpcs -fallow-half-arguments-and-returns -dwarf-column-info -debugger-tuning=lldb -target-linker-version 609.8 -v -resource-dir /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I ./src -I/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -internal-externc-isystem /Library/Developer/CommandLineTools/usr/include -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -fdebug-compilation-dir /Users/theo/Git/emacs-tree-sitter/langs/repos/c-sharp -ferror-limit 19 -fmessage-length 151 -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fobjc-runtime=macosx-11.0.0 -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/0l/7j7ds92x343fn8xnlrg1lv180000gp/T/parser-8e0edc.o -x c src/parser.c
clang -cc1 version 12.0.0 (clang-1200.0.32.29) default target arm64-apple-darwin20.3.0
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 ./src
 /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -no_deduplicate -dynamic -dylib -arch arm64 -platform_version macos 11.0.0 11.1 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o parser.so -L/usr/local/lib /var/folders/0l/7j7ds92x343fn8xnlrg1lv180000gp/T/parser-8e0edc.o -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture arm64:
  "_tree_sitter_c_sharp_external_scanner_create", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_deserialize", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_destroy", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_scan", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_serialize", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by theothornhill** [comment]
What remains from here for M1 support in MELPA? It is possible to generate bindings just fine, it seems :)

**Comment by ubolonton** [comment]
> Now I got it to work using this recipe:

Awesome!

>     6\. gcc -o parser.so --shared src/parser.c src/scanner.c -I./src

That's weird. IIRC, this is what `tree-sitter test` does underneath, if the shared lib was not compiled and put in place (`path/to/langs/bin/c_sharp.so`)

> What remains from here for M1 support in MELPA? It is possible to generate bindings just fine, it seems :)

We'd need to update CI jobs to build M1 binaries, and update the code (`tsc-dyn-get` and `tree-sitter-langs-build`) to handle both M1 and Intel chips.

However, since GitHub Actions and Azure Pipelines don't support M1 at the moment, what we can do now is just updating the docs with M1 installation instructions.

**Comment by lynlevenick** [comment]
> That's weird. IIRC, this is what tree-sitter test does underneath, if the shared lib was not compiled and put in place (path/to/langs/bin/c_sharp.so)

I just went through getting emacs-tree-sitter set up on an M1 laptop, and the `tree-sitter` I installed through `npm install --global tree-sitter-cli` was the x86_64 version. Similarly, `cargo install tree-sitter` got an (old and) x86_64 executable. Either of them, when running `tree-sitter test`, would invoke the x86_64 version of the compiler and create a x86_64 dylib, rather than an arm64 one. I had to clone the tree-sitter repo then `cd cli && cargo install --path .` to install an up-to-date correct architecture version.

tl;dr: for any other M1 users treading this path:

I had to `brew install cask llvm node rust` for dependencies.
My full `use-package` + `straight.el` invocation looks like this, though I can't guarantee _at all_ that this will work for you:

```elisp
(use-package tsc
  :straight `(:pre-build ,(when (and (memq window-system '(mac ns))
                                     (string-match-p (rx string-start "arm-")
                                                     system-configuration))
                                         ;; required for tree-sitter
                            (unless (and (executable-find "cargo")
                                         ;; required for building bindings
                                         (executable-find "cask")
                                         (executable-find "git")
                                         ;; required for tree-sitter to generate
                                         (executable-find "npm")
                                         ;; required for bindings
                                         (executable-find "llvm-gcc"))
                              (warn "tree-sitter build will fail"))
                              ;; get the latest tree-sitter
                            '(("sh" "-c" "test -d rust-tree-sitter || git clone https://github.com/tree-sitter/tree-sitter rust-tree-sitter; cd rust-tree-sitter && git pull")
                              ("sh" "-c" "cd rust-tree-sitter/cli && cargo install --path .")
                              ;; rebuild bindings
                              ("sh" "-c" "EMACS=emacs ./bin/setup && EMACS=emacs ./bin/build")
                              ;; ensure all language definitions
                              ("find" "langs/repos" "-type" "f" "-name" "grammar.js" "-not" "-path" "\\*/node_modules/\\*" "-exec" "sh" "-c" "grammar_path=\"${1%/*}\"; EMACS=emacs make \"ensure/${grammar_path##*/}\"" "sh" "{}" ";")
                              ;; needed or it will download x86_64 dylibs over the arm64 ones we just built
                              ("sh" "-c" "printf LOCAL >core/DYN-VERSION")))
                         :files ("core/DYN-VERSION" "core/tsc-dyn.*" "core/*.el")))
(use-package tree-sitter
  :commands (tree-sitter-hl-mode))
(use-package tree-sitter-langs
  ;; Don't clone the separate tree-sitter-langs repo, use the dylibs we
  ;; already built
  :straight (:host github :repo "ubolonton/emacs-tree-sitter"
             :files ("langs/*.el" ("bin" "langs/bin/*.dylib") ("queries" "langs/queries/*")))
  :after tree-sitter
  ;; If this isn't set then it'll download x86_64 dylibs over the arm64
  ;; dylibs we built
  :init (setf tree-sitter-langs--testing t))
```

**Comment by brurucy** [comment]
Still can't get it to work on my M1. @lynlevenick Could you please write a short walkthrough on how to do it manually?

**Comment by evolify** [comment]
Any update?

**Comment by ubolonton** [comment]
The binaries are available starting from [release 0.16.1](https://github.com/emacs-tree-sitter/elisp-tree-sitter/releases/tag/0.16.1).

Notes:
- The downloading code hasn't been updated, so you will have to manually download `tsc-dyn.aarch64-apple-darwin.dylib`, rename it to `tsc-dyn.dylib`, and put it next to `tsc.el`.
- It has not been tested. (Neither I nor GitHub Actions has M1).
- `tree-sitter-langs` hasn't provided M1 binaries yet.

**Comment by ubolonton** [comment]
Pre-compiled grammars for Apple Silicon are available starting from [tree-sitter-langs 0.10.13](https://github.com/emacs-tree-sitter/tree-sitter-langs/releases/tag/0.10.13).

Notes:
- The downloading code hasn't been updated, so you will have to manually download `tree-sitter-grammars.aarch64-apple-darwin.v0.10.13.tar.gz`, and extract the contents into `(tree-sitter-langs--bin-dir)`.
- They have not been tested. (Neither I nor GitHub Actions has M1).

**Comment by timlod** [comment]
Just reporting that I tried these now (hadn't been using Emacs on my M1 in a bit) and it seems to work fine so far.

**Comment by CadeMichael** [comment]
@timlod what version of emacs are you using? I'm using emacs-mac with use pacakge and am getting the error 

```
Error (use-package): tree-sitter/:catch: Cannot open load file: No such file or directory, tsc-dyn
Error (use-package): tree-sitter-langs/:catch: Cannot open load file: No such file or directory, tsc-dyn
```

i tried using straight-use-package with the same result.

**Comment by timlod** [comment]
See the last two messages by ubolonton - you need to manually download the files still and extract them into the tree-sitter directories.

**Comment by CadeMichael** [comment]
Thank you.
